nomad/resource_acl_auth_method.go

```diff
@@ -300,21 +429,44 @@ func fetchACLAuthMethodResource(client *api.Client, name string) (*api.ACLAuthMe
 	return authMethod, nil
 }
 
+// unredactACLAuthMethodResource mutates fetchedAuthMethod with real secrets,
+// so the full secret gets stored in TF state for diff checking, despite being
+// redacted in Nomad API responses.
 func unredactACLAuthMethodResource(d *schema.ResourceData, fetchedAuthMethod, originalAuthMethod *api.ACLAuthMethod) {
-	if originalAuthMethod != nil &&
-		originalAuthMethod.Config != nil &&
-		originalAuthMethod.Config.OIDCClientSecret != "" {
-		if fetchedAuthMethod.Config != nil {
+	// fetchedAuthMethod is what we retrieved from Nomad during a refresh, or after create/update.
+	// we will overwrite values on it, then that will be saved to TF state.
+	if fetchedAuthMethod == nil || fetchedAuthMethod.Config == nil {
+		return
+	}
+
+	// originalAuthMethod is what we just sent to Nomad, during resourceACLAuthMethod(Create|Update)
+	if originalAuthMethod != nil && originalAuthMethod.Config != nil {
+		if originalAuthMethod.Config.OIDCClientSecret != "" {
 			fetchedAuthMethod.Config.OIDCClientSecret = originalAuthMethod.Config.OIDCClientSecret
-			return
 		}
+		if originalAuthMethod.Config.OIDCClientAssertion != nil &&
+			originalAuthMethod.Config.OIDCClientAssertion.PrivateKey != nil &&
+			originalAuthMethod.Config.OIDCClientAssertion.PrivateKey.PemKey != "" {
+			fetchedAuthMethod.Config.OIDCClientAssertion.PrivateKey.PemKey = originalAuthMethod.Config.OIDCClientAssertion.PrivateKey.PemKey
```

Comment:
is `fetchedAuthMethod.Config.OIDCClientAssertion` guaranteed to be non-nil if `fetchedAuthMethod.Config` is? It doesn't look like it based on the nil check we do further down.

Reply:
anecdotally, it doesn't get tripped in any tests, but I still like to guard against `nil`, so I'll do some extra checkin :+1:

Reply:
the pile of conditions was feeling like a jenga tower, so I made a (pretty precious) wrapper to make this stretch of code easier to read (and safe). I could be convinced to pull back from that, but it does the thing. whatcha think?

Reply:
I'd have probably made those free functions but here in Terraform provider code it's not like we're instantiating this tramp object a million times in a server or something, so that's just a matter of taste, not function. LGTM